File: apps/backend-mock-template/utils/security/nonce-store.ts
/**
 * 内存 Nonce 存储：TTL 内同一 X-Request-ID 第二次 tryAcquire 返回 false。
 */

export interface NonceStore {
  tryAcquire(nonce: string, ttlMs: number): boolean;
  clear(): void;
}

export class MemoryNonceStore implements NonceStore {
  private readonly map = new Map<string, number>();

  tryAcquire(nonce: string, ttlMs: number): boolean {
    this.evictExpired();
    const now = Date.now();
    const existing = this.map.get(nonce);
    if (existing !== undefined && existing > now) {
      return false;
    }
    this.map.set(nonce, now + Math.max(1, ttlMs));
    return true;
  }

  clear(): void {
    this.map.clear();
  }

  private evictExpired(): void {
    const now = Date.now();
    for (const [k, exp] of this.map) {
      if (exp <= now) this.map.delete(k);
    }
  }
}

/** mock 进程内单例。 */
export const globalNonceStore = new MemoryNonceStore();

File: backend/java-admin/java-admin-infra/src/main/java/com/wshake/infra/security/NonceStore.java
package com.wshake.infra.security;

/**
 * Nonce 存储：在 TTL 内只允许同一 nonce 首次出现。
 *
 * @author wshake
 */
public interface NonceStore {

    /**
     * 尝试占用 nonce。
     *
     * @param nonce nonce 值（通常为 X-Request-ID）
     * @param ttlMs 过期时间（毫秒）
     * @return {@code true} 首次占用成功；{@code false} 已存在（重放）
     */
    boolean tryAcquire(String nonce, long ttlMs);
}
